### ISSUE TITLE
[BUG] SpeculativeDecodingTests fail sometimes

### ISSUE THREAD

**Issue by davidkoski** [ml-explore/mlx-swift-lm#353]
See #326 #315 and #323 plus recent failures in #344 

https://github.com/ml-explore/mlx-swift-lm/actions/runs/27645347325/job/81756903393?pr=344

```
◇ Test case passing 2 arguments numDraftTokens → 2, withLogitProcessor → false to "Speculative decoding matches default token generation" started.
✘ Test "Speculative decoding matches default token generation" recorded an issue with 2 arguments numDraftTokens → 2, withLogitProcessor → false at SpeculativeDecodingTests.swift:102:9: Expectation failed: (normalTokens → [39, 96, 93, 39, 39, 39, 39, 39, 39, 39, 37, 24, 24, 24, 17, 1, 77, 61, 59, 4, 1, 24, 61, 37, 99, 61, 37, 24, 17, 37, 24, 17]) == (speculativeTokens → [39, 37, 72, 37, 72, 37, 72, 96, 72, 96, 72, 59, 1, 65, 73, 1, 99, 24, 61, 59, 39, 41, 41, 24, 61, 59, 13, 64, 56, 38, 24, 61])
± inserted [93, 39, 39, 39, 39, 39, 39, 37, 24, 24, 17, 1, 77, 4, 1, 61, 37, 99, 37, 24, 17, 37, 24, 17], removed [37, 72, 37, 72, 37, 72, 96, 72, 72, 59, 1, 65, 73, 1, 99, 24, 61, 59, 41, 41, 13, 64, 56, 38]
```

I can get this to happen reliably by running the test in a loop -- it takes 2 or 3 iterations on my machine.

- `MLX_ENABLE_TF32=0` does not affect it (it shouldn't in #326 we switched to float16 to avoid this issue)
- curiously fixing the random seed also does not fix it (it doesn't prevent it or make it happen reliably)

**Comment by davidkoski** [comment]
@petrukha-ivan FYI in case you have any ideas -- I only looked into it briefly

**Comment by davidkoski** [comment]
I went back to e6a753aa9b42cb1a2fb9736e99ed5a4a9f40fb2e (the commit from #326) and if you run the tests in a loop it does fail there.

I think the next steps will be to re-read the commentary in #315 and see if I understand where this breaks.

**Comment by aleroot** [comment]
I think the root cause is that the test assumes token-by-token greedy decoding is bit-exact equivalent to batched speculative verification. That does not reliably hold across MLX/Metal execution paths.

Normal generation evaluates one token at a time (`L == 1`), while speculative decoding verifies `[current + draft tokens]` in one batched forward (`L == numDraftTokens + 1`). Those paths can use different attention/GEMM kernels, and when the top logits are very close, argmax can flip. Once the verifier accepts according to the batched logits, the generated prefix diverges from the normal token-by-token path and the rest snowballs.

#326 fixed the float32/TF32 version of this by moving the test model to float16, but it doesn’t remove the deeper assumption that batched and single-token logits are identical. I’d change this test to use a deterministic fake/tiny model with large logit margins for exact equality, and keep the real Gemma3 random-model test as a smoke test rather than asserting full token equality.
